lib/signs/utilities/messages.ex

```diff
@@ -23,252 +26,287 @@ defmodule Signs.Utilities.Messages do
   end
 
-  @spec get_paging_headway_or_alert_messages(
-          Signs.Realtime.t(),
+  defp paginate({first_top, first_bottom}, {second_top, second_bottom}) do
+    {%Content.Message.GenericPaging{messages: [first_top, second_top]},
+     %Content.Message.GenericPaging{messages: [first_bottom, second_bottom]}}
+  end
+
+  defp filter_alert_status(:shuttles_transfer_station, :temporary_terminal), do: :none
+  defp filter_alert_status(:suspension_transfer_station, :temporary_terminal), do: :none
+  defp filter_alert_status(alert_status, _), do: alert_status
+
+  @spec filter_predictions(
+          [Predictions.Prediction.t()],
+          Signs.Utilities.SourceConfig.config(),
+          Engine.Config.sign_config(),
           DateTime.t(),
-          Engine.Alerts.Fetcher.stop_status(),
-          :top | :bottom
-        ) :: Signs.Realtime.line_content()
-  defp get_paging_headway_or_alert_messages(sign, current_time, alert_status, pos) do
-    config_and_alert_status =
-      case {sign, alert_status, pos} do
-        {%Signs.Realtime{source_config: {config, _bottom}}, {alert_status, _}, :top} ->
-          {config, alert_status}
-
-        {%Signs.Realtime{source_config: {config, _bottom}}, alert_status, :top} ->
-          {config, alert_status}
-
-        {%Signs.Realtime{source_config: {_top, config}}, {_, alert_status}, :bottom} ->
-          {config, alert_status}
-
-        {%Signs.Realtime{source_config: {_top, config}}, alert_status, :bottom} ->
-          {config, alert_status}
-
-        _ ->
-          nil
+          DateTime.t()
+        ) :: [Predictions.Prediction.t()]
+  defp filter_predictions(predictions, config, sign_config, current_time, scheduled) do
+    predictions
+    |> then(fn predictions -> if(sign_config == :headway, do: [], else: predictions) end)
+    |> then(fn predictions ->
+      case {in_early_am?(current_time, scheduled),
+            before_early_am_threshold?(current_time, scheduled)} do
+        {false, _} ->
+          predictions
+
+        {true, true} ->
+          # More than 40 minutes before the first scheduled trip, filter out all predictions.
+          []
+
+        {true, false} ->
+          # Less than 40 minutes before the first scheduled trip, filter out reverse predictions,
+          # except for Prudential or Symphony EB
+          Enum.reject(
+            predictions,
+            &(Signs.Utilities.Predictions.prediction_certainty(&1, config) ==
+                @reverse_prediction_certainty and
+                &1.stop_id not in ["70240", "70242"])
+          )
       end
+    end)
+  end
 
-    case config_and_alert_status do
-      {config, alert_status} ->
-        get_paging_alert_message(alert_status, sign.uses_shuttles, config.headway_destination) ||
-          Signs.Utilities.Headways.get_paging_message(sign, config, current_time)
+  defp in_early_am?(current_time, scheduled) do
+    Timex.between?(DateTime.to_time(current_time), @early_am_start, DateTime.to_time(scheduled))
```

Comment:
I'm seeing all of the sign processes repeatedly fail on startup due to `scheduled` here being `nil` since the ETS table takes a second to get populated which seems like it's putting the application into a restart loop. In the old early AM code we just had some defaults for when the first scheduled departure is `nil` so maybe we just need to handle that in the new code as well. Are you seeing the same behavior?

Reply:
Oops, yes, I'm seeing that too. I think I changed this late in the refactor and forgot to recheck the with an actual run. I'll add some nil handling.